File: 4-estruturas-e-bibliotecas/1451-(tle).c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


struct node {
    char *text;
    struct node *next;
};

struct node *createNode(struct node * next);


int main() {
    char text[100001] = {""};

    while(scanf("%s", text) != EOF) {
        struct node *linkedList, *head, *tail;
        char aux[100001] = {""};
        int size = strlen(text), i=0, index=0;

        linkedList = createNode(NULL);
        linkedList->next = createNode(NULL);
        head = linkedList;
        tail = linkedList->next;
        for(; i<size; i++) {
            if(text[i] == '[') {
                strncpy(linkedList->text, aux, index);
                head = createNode(head);
                linkedList = head;

                strcpy(aux, "");
                index = 0;
                continue;

            }else if(text[i] == ']') {
                strncpy(linkedList->text, aux, index);
                tail->next = createNode(NULL);
                tail = tail->next;
                linkedList = tail;

                strcpy(aux, "");
                index = 0;
                continue;
            }

            aux[index] = text[i];
            index++;
        }

        strncpy(linkedList->text, aux, index);
        while(head) {
            printf("%s", head->text);
            head = head->next;
        }
    }

    return 0;
}


struct node *createNode(struct node * next) {
    struct node *newNode;

    newNode = (struct node *) malloc(sizeof(struct node *));
    newNode->text = (char *) malloc(sizeof(char) * 100001);
    newNode->next = next;

    return newNode;
}

File: 4-estruturas-e-bibliotecas/1451-(tle).py
class Node:
  def __init__(self, text=None, next=None, previous=None):
    self.text = text
    self.next = next
    self.previous = previous

#   def __str__(self):
#     return str(self.text)


while True:
    try:
        text = input()

        aux = ''
        linked_list = Node('')
        for c in text:
            if c == '[':
                while linked_list.previous:
                    linked_list = linked_list.previous
                
                linked_list = Node('', linked_list)
                linked_list.next.previous = linked_list
                aux = ''
                continue

            elif c == ']':
                while linked_list.next:
                    linked_list = linked_list.next
                
                linked_list.next = Node('', None, linked_list)
                linked_list = linked_list.next
                aux = ''
                continue
            
            aux = aux + c
            linked_list.text = aux
        
        # Go back to the start
        while linked_list.previous:
            linked_list = linked_list.previous
        
        while linked_list:
            print(linked_list.text, end='')
            linked_list = linked_list.next
        
        print('')

    except EOFError:
        break